### ISSUE TITLE
Declarations in wrong specificity order

### ISSUE THREAD

**Issue by jedmao** [postcss/postcss-nested#92]
## Steps to reproduce

Run [`postcss-cli`](https://github.com/postcss/postcss-cli) with the following command:

```zsh
postcss test.css -u postcss-nested
```

Given the following input from `test.css`:

```css
.a {
  .b {
    c: d;
    &e {
      f: g
    }
    h: i
  }
}
```

I am given the following output:

```css
.a .b {
    c: d;
    h: i;
  }
    .a .be {
      f: g
    }
```

## Issues

1. Both rules `.a .b` and `.a .be` have the same specificity, so the `h: i` declaration should come at the very end in a new `.a .b` rule, based on the order in which declarations were originally defined (not [how Sass does it][], but more accurate).
1. ~The `h: i` declaration was given a semi-colon that it didn't previously have. This might be an issue with the structure of the AST `raws`, as the rule defines the semi-colon and not the declaration?~
1. ~All rule indentation should be at the root of the document (definitely [how Sass does it][]).~
1. ~Sending the AST directly through `postcss-nested` doesn't completely unwrap the rules. I have to call the plugin multiple times to get complete unwrapping. If I have rules nested 4 levels deep, I have to call the plugin 4x!~

For the last issue, this is what I was doing from within a custom plugin:

```ts
const nested = postcssNested();
return (root, result) => {
  // do stuff
  nested(root, result);
};
```

Also, hi @ai! 👋 Long time, no see!

[how sass does it]: https://www.sassmeister.com/gist/1617eeb5b8775413d03650e520b06bec

**Comment by ai** [comment]
> 1. Both rules .a .b and .a .be have the same specificity, so the h: i declaration should come at the very end in a new .a .b rule, based on the order in which declarations were originally defined (not how Sass does it, but more accurate).

Yeap, let’s fix it. Can you send PR, because I am working on another project right now?

> 2. The h: i declaration was given a semi-colon that it didn't previously have. 
> 3. All rule indentation should be at the root of the document (definitely how Sass does it).

I decided that `postcss-nested` will not take care bout good formation. It is hard and most of the users will even not see this.

It could increase code 2x-3x times, especially if we will take care of all use cases.

> 4. Sending the AST directly through postcss-nested doesn't completely unwrap the rules. I have to call the plugin multiple times to get complete unwrapping. If I have rules nested 4 levels deep, I have to call the plugin 4x!

It is very strange because PostCSS itself does exactly the same `nested(root, result)` call.

Can you create a simple test stand to reproduce the issue?

> Also, hi @ai! wave Long time, no see!

Nice to see you again =^_^=. I started to use TypeScript in some of my projects. Also, I am planning to visit Austin in August.

**Comment by jedmao** [comment]
1. I can't, actually.
1. In hindsight, I agree.
1. Agreed.
1. I figured out that the culprit was my plugin when I converted at-rules into rules w/o using `newRule.append`. Turns out you also can't loop through `container.nodes`, which is what I was doing as well. Would love a generator method so I can do `for (const node of rule.nodes())`, because then I can `continue` and `break` and even `yield` if I wanted to.

I sent you an email with my phone number for when you come to Austin 😃

**Comment by ai** [comment]
@jedmao I created a task https://cultofmartians.com/tasks/postcss-nested.html#task

**Comment by jedmao** [comment]
Excellent! Thanks for doing that!

**Comment by demikhovr** [comment]
https://github.com/postcss/postcss-nested/pull/93

**Comment by ai** [comment]
The fix by @demikhovr was released in 4.2.2
